fun.lua

```diff
@@ -1032,6 +1040,86 @@ methods.operator = operator
 exports.op = operator
 methods.op = operator
 
+--------------------------------------------------------------------------------
+-- Stateful iterators
+--------------------------------------------------------------------------------
+
+-- Returns 'gen' throwing an error that explains that method is not supported
+local unsupported_stateful_method_gen = function(method)
+    return function()
+        error(string.format("stateful iterators do not support method '%s'",
+              method))
+    end
+end
+
+-- Helper of 'wrap_with_elems'
+local gen_with_elems_wrapper = function(gen_param_elems, state)
+    local elems = gen_param_elems[3]
+    if not elems then
+        return gen_param_elems[1](gen_param_elems[2], state)
+    else
+        gen_param_elems[3] = nil
+        return state, unpack(elems)
+    end
+end
+
+-- Wrap an iterator that yields all values from 'elems' table on the first
+-- iteration and then works in the same way as 'wrap(gen, param, state)'
+local wrap_with_elems = function(elems, gen, param, state)
+    local gen_param_elems = {gen, param, elems}
+    return wrap(gen_with_elems_wrapper, gen_param_elems, state)
+end
+
+-- Checks if drop_while should continue skipping. If iterator is not exhausted,
+-- elements returned by iterator are wrapped into a table and returned as the
+-- third return value
+local stateful_drop_while_check = function(fun, state_x, ...)
+    if state_x == nil then
+        return state_x, false
+    end
+    return state_x, fun(...), {...}
+end
+
+-- Implementation of method drop_while: original one copies iterator state to
+-- iterate over the same element twice - it doesn't work with stateful iterators
+local stateful_drop_while = function(fun, gen_x, param_x, state_x)
+    assert(type(fun) == "function", "invalid first argument to drop_while")
+    local cont = true
+    local elems
+    while state_x ~= nil and cont do
+        state_x, cont, elems = stateful_drop_while_check(fun, gen_x(param_x, state_x))
+    end
+    if state_x == nil then
+        return wrap(nil_gen, nil, nil)
+    end
+    return wrap_with_elems(elems, gen_x, param_x, state_x)
+end
+
+local stateful_methods = deepcopy(methods)
+-- Method 'is_null' cannot be supported since it advances iterator and
+-- does not return any new iterator, so passed stateful iterator is advanced
+-- and returned value is lost
+stateful_methods.is_null = unsupported_stateful_method_gen('is_null')
+-- Method 'cycle' is not supported because it would need to save all returned
+-- values since stateful iterator cannot be rewind to the beginning
+stateful_methods.cycle = unsupported_stateful_method_gen('cycle')
```

Comment:
If we will decide to replace the existing `drop_while` implementation with the new one that supports a stateful iterator, there are much less reasons to introduce the new iterator type.

In assumption that `drop_while` is reimplemented, can we leave it for a user to verify that `is_null` and `cycle` produce expected results and don't introduce the stateless iterator?

Reply:
I replaced `drop_while` with the new implementation and dropped stateful iterators completely.